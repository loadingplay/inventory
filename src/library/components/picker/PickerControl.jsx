import React from 'react';
import PropTypes from "prop-types";
import { PickerStore } from '../../services';
import "./picker.css";

export class PickerControl extends React.Component {

  static propTypes = {
    cellar_id: PropTypes.number,
    item: PropTypes.object
  }

  constructor(props) {
    super(props);
    this.state = {
      q: PickerStore.get(this.props.cellar_id, this.props.item.sku),
      options: {}
    };

    if (this.props.item.hq_inventory === undefined || this.props.item.current_inventory === undefined)
      this.state.options.disabled = 'disabled';
  }

  componentDidUpdate = () => {
    if (this.state.options.disabled !== undefined
      && this.props.item.hq_inventory !== undefined
      && this.props.item.current_inventory !== undefined
    ) {
      this.setState({ options: {} });
    }
  }

  handleUpdate = () => {
    PickerStore.set(
      this.props.cellar_id,
      this.props.item.sku,
      this.state.q,
      this.props.item.hq_inventory,
      this.props.item.current_inventory,
      this.props.item.suggested
    );
  }

  handlePlus = () => {
    this.setState({q: this.state.q + 1}, this.handleUpdate);
  }

  handleMinus = () => {
    if (this.state.q <= 0) return;
    this.setState({q: this.state.q - 1}, this.handleUpdate);
  }

  handleChange = (event) => {
    let new_value = parseInt(event.target.value);
    if (isNaN(new_value)) return;
    this.setState({q: new_value}, this.handleUpdate);
  }

  render = () => {
    return (
      <div className="btn-toolbar" role="toolbar" aria-label="Toolbar with button groups">
        <div className="btn-group" role="group" aria-label="First group">
          <button {...this.state.options} type="button" onClick={this.handleMinus} className="btn btn-outline-secondary">-</button>
          <input {...this.state.options} type="text" className="form-control text-center picker-input" onChange={this.handleChange} value={this.state.q} />
          <button {...this.state.options} type="button" onClick={this.handlePlus} className="btn btn-outline-secondary">+</button>
        </div>
      </div>
    )
  }
}
