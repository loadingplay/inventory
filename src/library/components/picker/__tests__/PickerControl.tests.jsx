import React from 'react';
import { shallow } from "enzyme";

import { PickerControl } from "../PickerControl";

describe("PickerControl", () => {
  test("it should render properly", () => {
    const wrapper = shallow(
      <PickerControl
        item={{
          sku: "test-sku"
        }}
      ></PickerControl>
    );
    expect(wrapper).toMatchSnapshot("detault");
    expect(wrapper.instance().state.options.disabled).toEqual('disabled');
  });

  test("it should update options when it is initialized with all data", () => {
    const wrapper = shallow(
      <PickerControl
        item={{
          sku: "test-sku",
          hq_inventory: 10,
          current_inventory: 5
        }}
      >
      </PickerControl>
    );

    expect(wrapper.instance().state.options).toEqual({})
  });

  test("it should add and substract 1 from picker", async () => {
    const wrapper = shallow(
      <PickerControl
        item={{
          sku: "test-sku",
          hq_inventory: 10,
          current_inventory: 5
        }}
      ></PickerControl>
    );

    expect(wrapper.instance().state.q).toEqual(0);
    wrapper.find("button").at(1).simulate('click');
    expect(wrapper.instance().state.q).toEqual(1);
    wrapper.find("button").at(0).simulate('click');
    expect(wrapper.instance().state.q).toEqual(0);
  });
});
