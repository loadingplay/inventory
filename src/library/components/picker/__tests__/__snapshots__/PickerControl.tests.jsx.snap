// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerControl it should render properly: detault 1`] = `
<div
  aria-label="Toolbar with button groups"
  className="btn-toolbar"
  role="toolbar"
>
  <div
    aria-label="First group"
    className="btn-group"
    role="group"
  >
    <button
      className="btn btn-outline-secondary"
      disabled="disabled"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
    <input
      className="form-control text-center picker-input"
      disabled="disabled"
      onChange={[Function]}
      type="text"
      value={0}
    />
    <button
      className="btn btn-outline-secondary"
      disabled="disabled"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
  </div>
</div>
`;
